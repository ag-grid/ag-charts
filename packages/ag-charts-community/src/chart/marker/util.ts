import { Circle } from './circle';
import { Cross } from './cross';
import { Diamond } from './diamond';
import { Heart } from './heart';
import type { Marker } from './marker';
import { Plus } from './plus';
import { Square } from './square';
import { Triangle } from './triangle';

type MarkerConstructor = new () => Marker;
type MarkerSupportedShapes = 'circle' | 'cross' | 'diamond' | 'heart' | 'plus' | 'square' | 'triangle';
export type MarkerShape = MarkerConstructor | MarkerSupportedShapes;

const MARKER_SHAPES: { [K in MarkerSupportedShapes]: MarkerConstructor } = {
    circle: Circle,
    cross: Cross,
    diamond: Diamond,
    heart: Heart,
    plus: Plus,
    square: Square,
    triangle: Triangle,
};

const MARKER_SUPPORTED_SHAPES = Object.keys(MARKER_SHAPES);

export function isMarkerShape(shape: MarkerShape | string): shape is MarkerSupportedShapes {
    return typeof shape === 'string' && MARKER_SUPPORTED_SHAPES.includes(shape);
}

// This function is in its own file because putting it into SeriesMarker makes the Legend
// suddenly aware of the series (it's an agnostic component), and putting it into Marker
// introduces circular dependencies.
export function getMarker(shape: MarkerShape | string = Square): MarkerConstructor {
    if (isMarkerShape(shape)) {
        return MARKER_SHAPES[shape];
    }
    if (typeof shape === 'function') {
        return shape;
    }
    return Square;
}
