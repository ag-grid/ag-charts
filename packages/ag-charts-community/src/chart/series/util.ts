import type { ChartAxis } from '../chartAxis';
import type { ProcessedDataDef } from '../data/dataModel';

export function convertValuesToScaleByDefs<T extends string>({
    defs,
    values,
    xAxis,
    yAxis,
}: {
    defs: [string, ProcessedDataDef[]][];
    values: Record<T, unknown>;
    xAxis: ChartAxis | undefined;
    yAxis: ChartAxis | undefined;
}): Record<T, number> {
    if (!(xAxis && yAxis)) {
        throw new Error('Axes must be defined');
    }
    const result: Record<string, number> = {};
    for (const [searchId, [{ def }]] of defs) {
        if (Object.hasOwn(values, searchId)) {
            const { scale } = def.type === 'key' ? xAxis : yAxis;
            result[searchId] = Math.round(scale.convert((values as any)[searchId]));
        }
    }
    return result;
}
