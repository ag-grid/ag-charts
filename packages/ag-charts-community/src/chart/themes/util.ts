import type { AgCartesianChartOptions } from 'ag-charts-types';

type CartesianAxis = Exclude<AgCartesianChartOptions['axes'], undefined>[0];

export function swapAxisCondition(axes: [CartesianAxis, CartesianAxis], swap: (series: any) => boolean) {
    return (series: any) => {
        if (swap(series)) {
            return [
                { ...axes[0], position: axes[1].position },
                { ...axes[1], position: axes[0].position },
            ];
        } else {
            return axes;
        }
    };
}
