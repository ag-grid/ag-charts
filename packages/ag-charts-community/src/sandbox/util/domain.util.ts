export function convertDomain(
    n: number,
    [d0, d1]: [number, number] | number[],
    [r0, r1]: [number, number] | number[],
    clamp?: boolean
) {
    if (d0 === d1) {
        return (r0 + r1) / 2;
    } else if (n === d0) {
        return r0;
    } else if (n === d1) {
        return r1;
    }

    if (clamp) {
        if (n < d0) {
            return r0;
        } else if (n > d1) {
            return r1;
        }
    }

    return r0 + ((n - d0) / (d1 - d0)) * (r1 - r0);
}

export function niceDomain(
    domain: number[],
    options: { floor: (n: number) => number; ceil: (n: number) => number } = Math
) {
    const d = [...domain];

    if (d.length === 0) {
        d.push(0, 1);
    } else if (d.length === 1) {
        d.push(d[0]);
    }

    const i = d.length - 1;
    d[0] = options.floor(d[0]);
    d[i] = options.ceil(d[i]);
    d[i] += 0; // add 0 to convert -0 into 0
    return d;
}
