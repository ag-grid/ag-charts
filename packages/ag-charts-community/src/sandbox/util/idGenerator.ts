// Generates sequential numeric IDs, resetting to 0 when the maximum safe integer is reached.
export class IdGenerator {
    private currentId: number = 0;

    /**
     * Generates the next ID, resetting to 0 if the maximum safe integer value is reached.
     * @returns {number} The next ID.
     */
    generateId(): number {
        if (this.currentId === Number.MAX_SAFE_INTEGER) {
            this.currentId = 0;
        }
        return this.currentId++;
    }

    // Resets the current ID counter to zero.
    reset() {
        this.currentId = 0;
    }
}
