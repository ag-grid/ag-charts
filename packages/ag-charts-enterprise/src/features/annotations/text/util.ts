import { type TextAlign, _ModuleSupport, _Scene } from 'ag-charts-community';

const { TextWrapper, CachedTextMeasurerPool } = _ModuleSupport;
const { BBox } = _Scene;

export type AnnotationTextPosition = 'top' | 'center' | 'bottom';
export type AnnotationTextAlignment = 'left' | 'center' | 'right';

export type TextOptions = _ModuleSupport.FontOptions & { textAlign: TextAlign; position: AnnotationTextPosition };

export const ANNOTATION_TEXT_LINE_HEIGHT = 1.38;

export function getTextWrapOptions(options: TextOptions): Omit<_ModuleSupport.WrapOptions, 'maxWidth'> {
    return {
        font: {
            fontFamily: options.fontFamily,
            fontSize: options.fontSize,
            fontStyle: options.fontStyle,
            fontWeight: options.fontWeight,
        },
        textAlign: options.textAlign,
        textBaseline: (options.position == 'center' ? 'middle' : options.position) as CanvasTextBaseline,
        lineHeight: ANNOTATION_TEXT_LINE_HEIGHT,
        avoidOrphans: false,
        textWrap: 'always',
    };
}

export function wrapText(options: TextOptions, text: string, width: number) {
    return width
        ? TextWrapper.wrapText(text, {
              ...getTextWrapOptions(options),
              maxWidth: width,
          })
        : text;
}

export function measureAnnotationText(options: TextOptions, text: string) {
    const textOptions = getTextWrapOptions(options);

    const { lineMetrics, width } = CachedTextMeasurerPool.measureLines(text, textOptions);
    const height = lineMetrics.length * (options.fontSize ?? 14) * ANNOTATION_TEXT_LINE_HEIGHT;

    return {
        width,
        height,
    };
}

export function getBBox(
    options: TextOptions & { width?: number },
    text: string,
    coords: _ModuleSupport.Vec2,
    bbox?: _Scene.BBox
) {
    let width = bbox?.width ?? 0;
    let height = bbox?.height ?? 0;

    if (!bbox) {
        const wrappedText = options.width != null ? wrapText(options, text, options.width) : text;
        ({ width, height } = measureAnnotationText(options, wrappedText));
    }

    return new BBox(coords.x, coords.y, width, height);
}

export function updateTextNode(
    node: _Scene.Text,
    text: string,
    isPlaceholder: boolean,
    config: TextOptions & { visible?: boolean; color?: string; getPlaceholderColor: () => string | undefined },
    { x, y }: _ModuleSupport.Vec2
) {
    const { visible = true, fontFamily, fontSize = 14, fontStyle, fontWeight, textAlign } = config;
    const lineHeight = fontSize * ANNOTATION_TEXT_LINE_HEIGHT;
    const textBaseline = config.position == 'center' ? 'middle' : config.position;
    const fill = isPlaceholder ? config.getPlaceholderColor() : config.color;

    node.setProperties({
        x,
        y,
        visible,
        text,
        fill,
        fontFamily,
        fontSize,
        fontStyle,
        fontWeight,
        textAlign,
        lineHeight,
        textBaseline,
    });
}
