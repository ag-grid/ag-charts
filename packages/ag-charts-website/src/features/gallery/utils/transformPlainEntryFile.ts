import j from 'jscodeshift';
import { filterPropertyKeys } from '@utils/jsCodeShiftUtils';
// eslint-disable-next-line @nx/enforce-module-boundaries
import { parseExampleOptions } from '../../../../../ag-charts-community/src/chart/test/load-example';

/**
 * JS Code Shift transformer to generate plain entry file
 */

function transformer(sourceFile: string, dataFile?: string) {
    const root = j(sourceFile);

    const optionsExpression = root
        .find(j.VariableDeclarator, {
            id: {
                type: 'Identifier',
                name: 'options',
            },
        })
        .find(j.ObjectExpression);

    // Find and remove properties in the 'options' object
    const propertiesToRemove = ['subtitle', 'footnote', 'legend'];
    optionsExpression.forEach((path) => {
        path.node.properties = filterPropertyKeys({
            removePropertyKeys: propertiesToRemove,
            properties: path.node.properties,
        });
    });

    // Add disabled legend
    const legendPropertyNode = j.property(
        'init',
        j.identifier('legend'),
        j.objectExpression([j.property('init', j.identifier('enabled'), j.literal(false))])
    );

    const optionsNode = optionsExpression.get(0).node;
    optionsNode.properties = optionsNode.properties.filter((property: any) => property.key?.name !== 'padding');
    const optionsExpressionProperties = optionsNode.properties;
    optionsExpressionProperties.push(legendPropertyNode);

    // Axes
    optionsExpression
        .find(j.Property, {
            key: {
                name: 'axes',
            },
        })
        .find(j.ObjectExpression)
        .forEach((path) => {
            const propertiesNode = path.node;

            // Remove axis title
            const title = propertiesNode.properties.find((prop) => prop.key.name === 'title');
            if (title) {
                propertiesNode.properties = filterPropertyKeys({
                    removePropertyKeys: ['title'],
                    properties: propertiesNode.properties,
                });
            }
            propertiesNode.properties.push(
                j.property(
                    'init',
                    j.identifier('title'),
                    j.objectExpression([j.property('init', j.identifier('enabled'), j.literal(false))])
                )
            );
        });

    // Padding
    const paddingPropertyNode = j.property(
        'init',
        j.identifier('padding'),
        j.objectExpression([
            j.property('init', j.identifier('top'), j.literal(10)),
            j.property('init', j.identifier('right'), j.literal(20)),
            j.property('init', j.identifier('bottom'), j.literal(10)),
            j.property('init', j.identifier('left'), j.literal(20)),
        ])
    );
    optionsExpressionProperties.push(paddingPropertyNode);

    const code = root.toSource();
    const options = parseExampleOptions('options', code, dataFile);

    return { code, options };
}

export function transformPlainEntryFile(entryFile: string, dataFile?: string): { code: string; options: {} } {
    return transformer(entryFile, dataFile);
}
