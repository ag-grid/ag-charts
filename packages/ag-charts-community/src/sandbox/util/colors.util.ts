import { createElement, getDocument } from '../../util/dom';
import { clamp } from '../../util/number';

export type RGBA = { r: number; g: number; b: number; a: number };
export type OKLCH = { l: number; c: number; h: number; a: number };

export function stringToRgba(colorName: string, container = getDocument('body')): RGBA {
    const span = container.appendChild(createElement('span', { color: colorName }));
    const { color } = getComputedStyle(span);
    const [r, g, b, a = 1] = color
        .slice(color.indexOf('(') + 1, -1)
        .split(',')
        .map(Number);
    getDocument('body').removeChild(span);
    return { r, g, b, a };
}

export function rgbaToOklch(r: number, g: number, b: number, a: number = 1): OKLCH {
    // Step 1: Convert RGB to linear sRGB
    const linearR = gammaCorrect(r / 255);
    const linearG = gammaCorrect(g / 255);
    const linearB = gammaCorrect(b / 255);

    // Step 2: Convert linear sRGB to LMS using the matrix
    const lmsL = Math.cbrt(0.4122214708 * linearR + 0.5363325363 * linearG + 0.0514459929 * linearB);
    const lmsM = Math.cbrt(0.2119034982 * linearR + 0.6806995451 * linearG + 0.1073969566 * linearB);
    const lmsS = Math.cbrt(0.0883024619 * linearR + 0.2817188376 * linearG + 0.6299787005 * linearB);

    // Step 3: Convert LMS to OKLab
    const okLabL = 0.2104542553 * lmsL + 0.793617785 * lmsM - 0.0040720468 * lmsS;
    const okLabA = 1.9779984951 * lmsL - 2.428592205 * lmsM + 0.4505937099 * lmsS;
    const okLabB = 0.0259040371 * lmsL + 0.7827717662 * lmsM - 0.808675766 * lmsS;

    // Step 4: Convert OKLab to OKLCH
    const chroma = Math.hypot(okLabA, okLabB);
    const hue = (Math.atan2(okLabB, okLabA) * 180) / Math.PI;

    return { l: okLabL, c: chroma, h: hue >= 0 ? hue : hue + 360, a };
}

export function oklchToRgba(l: number, c: number, h: number, a: number = 1): RGBA {
    // Step 1: Convert OKLCH to OKLab
    const okLabA = c * Math.cos((h * Math.PI) / 180);
    const okLabB = c * Math.sin((h * Math.PI) / 180);

    // Step 2: Convert OKLab to LMS
    const lmsL = (l + 0.3963377774 * okLabA + 0.2158037573 * okLabB) ** 3;
    const lmsM = (l - 0.1055613458 * okLabA - 0.0638541728 * okLabB) ** 3;
    const lmsS = (l - 0.0894841775 * okLabA - 1.291485548 * okLabB) ** 3;

    // Step 3: Convert LMS to linear sRGB
    const linearR = 4.0767416621 * lmsL - 3.3077115913 * lmsM + 0.2309699292 * lmsS;
    const linearG = -1.2684380046 * lmsL + 2.6097574011 * lmsM - 0.3413193965 * lmsS;
    const linearB = -0.0041960863 * lmsL - 0.7034186147 * lmsM + 1.707614701 * lmsS;

    // Step 4: Convert linear sRGB to RGB
    return {
        r: Math.round(clamp(0, gammaCorrect(linearR, true), 1) * 255),
        g: Math.round(clamp(0, gammaCorrect(linearG, true), 1) * 255),
        b: Math.round(clamp(0, gammaCorrect(linearB, true), 1) * 255),
        a,
    };
}

// Helper function for gamma correction between sRGB and linear RGB
function gammaCorrect(channel: number, invert?: boolean): number {
    if (invert) {
        return channel <= 0.0031308 ? 12.92 * channel : 1.055 * Math.pow(channel, 1 / 2.4) - 0.055;
    }
    return channel <= 0.04045 ? channel / 12.92 : Math.pow((channel + 0.055) / 1.055, 2.4);
}
