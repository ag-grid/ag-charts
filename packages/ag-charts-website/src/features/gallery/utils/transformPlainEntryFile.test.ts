import util from 'node:util';

import { transformPlainEntryFile } from './transformPlainEntryFile';

const getEntryFile = (chartOptions: object) => {
    // NOTE: Use `util.inspect` instead of `JSON.stringify` so that there aren't quotes around
    // object keys, just like actual source code
    const chartOptionsStr = util.inspect(chartOptions, { depth: 10 });
    const entryFile = `const options = ${chartOptionsStr};\nvar chart = agCharts.AgCharts.create(options);`;

    return entryFile;
};

const getChartsOptionsPlainEntryFile = (chartsOptions: object) => {
    const sourceStr = getEntryFile(chartsOptions);
    const output = transformPlainEntryFile(sourceStr);

    return output;
};

describe('transformPlainEntryFile', () => {
    test('object key as literal with strings ');

    test('default entry file', () => {
        expect(getChartsOptionsPlainEntryFile({}).code).toMatchSnapshot();
    });

    test('remove subtitle', () => {
        expect(
            getChartsOptionsPlainEntryFile({
                subtitle: {
                    enabled: true,
                    text: 'Test Subtitle',
                },
            }).code
        ).toMatchSnapshot();
    });

    test('remove footnote', () => {
        expect(
            getChartsOptionsPlainEntryFile({
                footnote: {
                    enabled: true,
                    text: 'Test footnote',
                },
            }).code
        ).toMatchSnapshot();
    });

    test('disable legend', () => {
        expect(
            getChartsOptionsPlainEntryFile({
                legend: {
                    enabled: true,
                    text: 'Test legend',
                },
            }).code
        ).toMatchSnapshot();
    });

    test('Remove axes titles', () => {
        expect(
            getChartsOptionsPlainEntryFile({
                axes: [
                    {
                        type: 'category',
                        position: 'bottom',
                        title: {
                            text: 'CATEGORY AXIS TITLE',
                        },
                    },
                    {
                        type: 'number',
                        position: 'left',
                        title: {
                            text: 'NUMBER AXIS TITLE',
                        },
                    },
                ],
            }).code
        ).toMatchSnapshot();
    });

    test('Adjust padding', () => {
        expect(
            getChartsOptionsPlainEntryFile({
                seriesArea: {
                    padding: {
                        left: 40,
                    },
                },
                padding: {
                    top: 10,
                    right: 20,
                    bottom: 30,
                    left: 40,
                },
            }).code
        ).toMatchSnapshot();
    });
});
