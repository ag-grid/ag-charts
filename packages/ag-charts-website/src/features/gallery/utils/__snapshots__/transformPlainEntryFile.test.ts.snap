// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformPlainEntryFile > Adjust padding 1`] = `
"const options = {
  seriesArea: { padding: { left: 40 } },

  legend: {
    enabled: false
  },

  gradientLegend: {
    enabled: false
  },

  theme: {
    baseTheme: \\"ag-default\\",

    overrides: {
      common: {
        axes: {
          category: {
            label: {
              autoRotate: false
            }
          }
        }
      }
    }
  },

  padding: {
    top: 10,
    right: 20,
    bottom: 10,
    left: 20
  }
};
var chart = agCharts.AgChart.create(options);"
`;

exports[`transformPlainEntryFile > Remove axes titles 1`] = `
"const options = {
  axes: [
    {
      type: 'category',
      position: 'bottom',
      title: {
        enabled: false
      }
    },
    {
      type: 'number',
      position: 'left',
      title: {
        enabled: false
      }
    }
  ],

  legend: {
    enabled: false
  },

  gradientLegend: {
    enabled: false
  },

  theme: {
    baseTheme: \\"ag-default\\",

    overrides: {
      common: {
        axes: {
          category: {
            label: {
              autoRotate: false
            }
          }
        }
      }
    }
  },

  padding: {
    top: 10,
    right: 20,
    bottom: 10,
    left: 20
  }
};
var chart = agCharts.AgChart.create(options);"
`;

exports[`transformPlainEntryFile > default entry file 1`] = `
"const options = {
  legend: {
    enabled: false
  },

  gradientLegend: {
    enabled: false
  },

  theme: {
    baseTheme: \\"ag-default\\",

    overrides: {
      common: {
        axes: {
          category: {
            label: {
              autoRotate: false
            }
          }
        }
      }
    }
  },

  padding: {
    top: 10,
    right: 20,
    bottom: 10,
    left: 20
  }
};
var chart = agCharts.AgChart.create(options);"
`;

exports[`transformPlainEntryFile > disable legend 1`] = `
"const options = {
  legend: {
    enabled: false
  },

  gradientLegend: {
    enabled: false
  },

  theme: {
    baseTheme: \\"ag-default\\",

    overrides: {
      common: {
        axes: {
          category: {
            label: {
              autoRotate: false
            }
          }
        }
      }
    }
  },

  padding: {
    top: 10,
    right: 20,
    bottom: 10,
    left: 20
  }
};
var chart = agCharts.AgChart.create(options);"
`;

exports[`transformPlainEntryFile > remove footnote 1`] = `
"const options = {
  legend: {
    enabled: false
  },

  gradientLegend: {
    enabled: false
  },

  theme: {
    baseTheme: \\"ag-default\\",

    overrides: {
      common: {
        axes: {
          category: {
            label: {
              autoRotate: false
            }
          }
        }
      }
    }
  },

  padding: {
    top: 10,
    right: 20,
    bottom: 10,
    left: 20
  }
};
var chart = agCharts.AgChart.create(options);"
`;

exports[`transformPlainEntryFile > remove subtitle 1`] = `
"const options = {
  legend: {
    enabled: false
  },

  gradientLegend: {
    enabled: false
  },

  theme: {
    baseTheme: \\"ag-default\\",

    overrides: {
      common: {
        axes: {
          category: {
            label: {
              autoRotate: false
            }
          }
        }
      }
    }
  },

  padding: {
    top: 10,
    right: 20,
    bottom: 10,
    left: 20
  }
};
var chart = agCharts.AgChart.create(options);"
`;
